File: cs461/pa2/nfa2dfa.cpp
// nfa2dfa.c++
// Mitchell Wade
// CS461 - Compilers

#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstdlib>
#include <sstream>
#include <math.h>
#include <string>
#include <vector>
#include <fstream>
#include <algorithm>
#include <map>
#include <cmath>
#include <stack>
#include <queue>

using namespace std;

typedef vector <int> IVec;
typedef vector <string> SVec;
typedef vector <double> DVec;

class NFA {
   public:
      vector <string> chars;

};


void output(int initialState, vector <int> finalStates, int totalStates, 
            vector <string> dfaSymbols) {

   int i = 0;

   printf("Final DFA:\n");
   printf("Initial State: %d\n", initialState);
   for (i = 0; i < finalStates.size(); i++) {
      if (finalStates.size() == 1) {
         printf("Final State: {");
         printf("%d", finalStates[i]);
         printf("}\n");
      }
      if (finalStates.size() >= 1) {
         if (i == 0) printf("Final States: {");
         printf("%d", finalStates[i]);
         if (i != finalStates.size()-1) printf(",");
         if (i == finalStates.size()-1) printf("}\n");
      }
   }
   printf("Total States: %d\n", totalStates);
   for (i = 0; i < dfaSymbols.size(); i++) {
      if (i == 0) printf("State: ");
      printf("%10s", dfaSymbols[i].c_str());
      if (i == dfaSymbols.size()-1) printf("\n");
   }
}

int main(int argc, char** argv) {

   int l = 0, w = 0, currentState = 0;
   int i = 0, j = 0, k = 0, initialState, totalStates;
   string s, state, symbol;
   istringstream ss;
   vector <int> finalStates;
   vector <string> nfaSymbols;
   vector <string> dfaSymbols;
   vector <int> tmpStates;
   vector <int> curStates;
   vector <string> tmpSymbols;
   map <string, int> transitions;
   map <string, int>::iterator tit;

   cout << "reading NFA ..." << endl;
   cout << endl;

   while (getline(cin, s)) {
      l++;
      ss.clear();
      ss.str(s);
      w = 0;
      while (ss >> s) {
         w++;
         //printf("Line %3d, word %3d: %s\n", l, w, s.c_str());

         // Create initial state
         if (s  == "Initial") {
            ss >> s;
            ss.clear();
            ss >> s;
            initialState = atoi(s.c_str());
         }

         // Create final states vector
         if (s == "Final") {
            ss >> s;
            ss.clear();
            ss >> s;
            for (i = 0; i < s.size(); i++) {
               if (s[i] == ',') {
                  finalStates.push_back(atoi(state.c_str()));
                  state = "";
                  
               }
               if (isdigit(s[i])) {
                  state += s[i];
               }
            }
            if (state != "") finalStates.push_back(atoi(state.c_str()));
         }

         // Create total states
         if (s == "Total") {
            ss >> s;
            ss.clear();
            ss >> s;
            totalStates = atoi(s.c_str());
         }

         // Create states
         if (s == "State") {
            while(ss >> s) {
               nfaSymbols.push_back(s.c_str());
               s.clear();
            }
         }

         // Create transitions
         currentState = atoi(s.c_str()); 
         printf("currentState = %d\n", currentState);
         i = 0;
         state = "";
         while(ss >> s) {
            printf("s: %s\n", s.c_str());
            if (s != "{}") {
               symbol = nfaSymbols[i];
               printf("symbol: %s\n", symbol.c_str());
               for (j = 0; j < s.size(); j++) {
                  if (s[j] != '{' && s[j] != ',' && s[j] != '}') {
                     state += s[j];
                     printf("state: %s\n", state.c_str());
                  }
                  if (s[j] == ',') {
                     transitions.insert(make_pair(symbol, atoi(state.c_str())));
                     tmpStates.push_back(atoi(state.c_str()));
                     tmpSymbols.push_back(symbol);
                     curStates.push_back(currentState);
                     printf("transition: %s\n", state.c_str());
                     state = "";
                  }
               }
               if (state != "") {
                  transitions.insert(make_pair(symbol, atoi(state.c_str())));
                  tmpStates.push_back(atoi(state.c_str()));
                  tmpSymbols.push_back(symbol);
                  curStates.push_back(currentState);
                  printf("transition: %s\n", state.c_str());
                  state = "";
               }
            }
               /*for (j = 0; j < s.size(); j++) {
                     if (s[j] == ',') {
                        tmpStates.push_back(atoi(state.c_str()));
                        transitions.insert(make_pair(symbol, atoi(state.c_str())));
                        state = "";
                     }
                     //if (isdigit(s[j])) {
                        state += s[j];
                     //}
               }*/
               /*if (state != "") {
                  tmpStates.push_back(atoi(state.c_str()));
                  transitions.insert(make_pair(symbol, atoi(state.c_str())));
                  state = "";
               }*/
            
            // Go to next symbol}
            i++;
         }

         /*for (i = 0; i < tmpStates.size(); i++) {
            transitions.insert(make_pair(symbol, atoi(state.c_str())));
         }*/
      }
   }
   
   // Create dfaSymbols vector
   for (i = 0; i < nfaSymbols.size(); i++) {
      if (nfaSymbols[i] != "E") dfaSymbols.push_back(nfaSymbols[i]);
   }

   cout << "creating corresponding DFA ..." << endl;
   cout << endl;

   output(initialState, finalStates, totalStates, dfaSymbols);

   printf("transitions size = %d\n", transitions.size());
   for(tit = transitions.begin(); tit != transitions.end(); tit++) {
      cout << "Symbol: " << tit->first << " State:" << tit->second << endl;
   }

   printf("tmpStates/tmpSymbol\n");
   for(i = 0; i < tmpStates.size(); i++) {
      printf("tmpStates/tmpSymbols/curStates[%d] = %d/%s/%d\n", i, tmpStates[i], tmpSymbols[i].c_str(), curStates[i]);
   }

   vector <int> NFA;
   NFA.resize(totalStates);
   /*for (i = 0; i < tmpStates.size(); i++) {
      NFA[i].resize(totalStates);
   }*/

   vector <int> DFA;

   // Epsilon Closure Function
   vector <int> states;
   queue <int> stateStack;
   int state1;
   vector <int> epsilonStates;
   vector <int> closure;

   states.push_back(initialState);
   
   for (i = 0; i < states.size(); i++) {
      stateStack.push(states[i]);
   }

   for (i = 0; i < states.size(); i++) {
      closure.push_back(states[i]);
   }

   while (!stateStack.empty()) {
      state1 = stateStack.front();
      stateStack.pop();
      for (i = 0; i < tmpStates.size(); i++) {
         if (tmpSymbols[i] == "E") {
            epsilonStates.push_back(tmpStates[i]);
         }
      }

      for (i = 0; i < epsilonStates.size(); i++) {
         for (i = 0; i < closure.size(); i++) {
            if (epsilonStates[i] != closure[i]) {
               //flag
               //check if it already exists
               closure.push_back(epsilonStates[i]);
               stateStack.push(epsilonStates[i]);
            }
         }
      }

      // Sort the closure list
      vector <int> finalClosure;
      //finalClosure = qsort(closure);

      // return finalClosure when inside epsilon function
   }


   //_dfa_init_state = Epsilon_Closure(states, tmpStates, tmpSymbols, curStates);
   
   return 0;
}

/*vector <int> Epsilon_Closure(vector <int> states, vector <int> tmpStates, vector <string> tmpSymbols, vector <int> curStates) {

   vector <int> finalClosure;
   stack <int> stateStack;

   for t in T:
      stateStack.
   return finalClosure;
}*/

File: cs461/pa2/nfa2dfa.py
# nfa2dfa.py
# Mitchell Wade
# CS461 - Compilers
# This program converts an NFA to its corresponding DFA using the construction
# algorithm.

import sys

# EPSILON CLOSURE FUNCTION
def epsilon_closure(T, NFA):
  stateStack = []
  
  for t in T:
    stateStack.append(t)

  closure = list(T)
  while stateStack:
    t = stateStack.pop()    #grab the next state
    epsilon_states = [x[1] for x in NFA[t] if x[0] == 'E'] #grab epsilon states
    print("epsilon_states\n")
    print(epsilon_states)
    for j in epsilon_states:
      if not j in closure:  #if its not there, add it
        closure.append(j)
        stateStack.append(j)

  finalClosure = tuple(sorted(closure))
  return finalClosure

# MOVE FUNCTION
def move(T, a, NFA):
  moveStates = set([])
  for n in T:
    stateIndex = [x[1] for x in NFA[n] if x[0] == a]
    moveStates.update(stateIndex)
  return tuple(moveStates)

# CREATES A NEW DFA STATE
def create_new_dfa_state(states, index, DFA, d2r, r2d):
  newState = index+1

  if r2d.has_key(states):
    print "error: raw DFA state already exists"
    print ("states: %s" % states), (" newState: %s" % newState)
    raise SystemExit(1)
  if d2r.has_key(newState):
    print "error: DFA state already exists"
    print ("states: %s" % states), (" newState: %s" % newState)
    raise SystemExit(1)

  r2d[states] = newState
  d2r[newState] = states
  DFA[newState] = []

  print "new DFA state: ", ("%d" % newState).ljust(3), " --> ", ("%s" % setstr(states))
  return newState

# CHECK STATE TABLE
def get_unmarked_state(DFA, mark):
  for dstate in DFA.keys():
    if not mark[dstate]:
      return dstate
  return None

# PRINTS NFA/DFA STATES AS SETS
def setstr(x):

  if not x:
    return "{}"

  ret = "{"
  for y in x[:-1]:
    ret += ("%d," % y)
  ret += "%d}" % x[-1]
  
  return ret

# MAIN FUNCTION
def nfa2dfa(args):

  print "reading NFA ...",
 
  # Read the NFA details from input file
  nfaInitialState = int(sys.stdin.readline().split()[2])
  nfaFinalStates = tuple([int(x) for x in sys.stdin.readline().split()[2].strip('{}').split(',')]) 
  nfaTotalStates = int(sys.stdin.readline().split()[2])
  nfaSymbols = sys.stdin.readline().split()[1:]
  dfaSymbols = [x for x in nfaSymbols if not x == 'E']

  NFA = {}
 
  # Create labeled pairs for the data
  for i in range(nfaTotalStates):  
    pts = sys.stdin.readline().split() 
    label = int(pts[0]) 
    transitions = [] 
    
    for j,tx in enumerate(pts[1:]): 
      if tx != '{}':
        s  = nfaSymbols[j] 
        tmp = set([int(x) for x in tx.strip('{}').split(',')])
        for x in tmp:
          transitions.append((s, x)) 
    
    NFA[label] = transitions 
  print "done."

  print "\ncreating corresponding DFA ..."
  DFA = {}
  d2r = {}
  r2d = {}
  currentStateDFA = 0
  print("NFA\n")
  print(NFA)
  print("\n")
  print("tuple([nfaInitialState])]\n")
  print tuple([nfaInitialState])
  print("\n")
  _dfa_init_state = epsilon_closure(tuple([nfaInitialState]), NFA)
  dfa_init_state = create_new_dfa_state(_dfa_init_state, currentStateDFA, DFA, d2r, r2d)
  currentStateDFA = dfa_init_state

  mark = {}
  mark[dfa_init_state] = False

  T = get_unmarked_state(DFA, mark)
  while T:
    mark[T] = True

    for sym in dfaSymbols:
      _U = epsilon_closure(move(d2r[T],sym,NFA), NFA)
      if _U:
        if not _U in r2d.keys():
          currentStateDFA = create_new_dfa_state(_U, currentStateDFA, DFA, d2r, r2d)
          mark[currentStateDFA] = False
        DFA[T].append((sym,r2d[_U]))

    T = get_unmarked_state(DFA, mark)

  dfa_final_states = tuple(sorted([x for x in DFA.keys()\
                       if any([ True if y in nfaFinalStates else False\
                              for y in d2r[x] ])]))
  print "done."
  print "\nfinal DFA:"

  auto = DFA
  init_state = dfa_init_state
  final_states = dfa_final_states
  symbols = dfaSymbols
  sorted_syms = sorted([x for x in symbols if not x == 'E'])
  if 'E' in symbols:
    sorted_syms.append('E')

  print "Initial State: ", ("%d" % init_state)
  print "Final States:  ", ("%s" % setstr(final_states))
  print "Total States:  ", ("%d" % len(auto.keys()))

  print "State".ljust(7),
  for sym in sorted_syms:
    print ("%s" % sym).ljust(16),
  print ""

  for state in sorted(auto.keys()):
    print ("%d" % state).ljust(7),
    for sym in sorted_syms:
      tmp = tuple(sorted([x[1] for x in auto[state] if x[0] == sym]))
      print ("%s" % setstr(tmp)).ljust(16),
    print ""

if __name__ == "__main__":
  nfa2dfa(sys.argv)
